tasks/egyptian-fractions/src/main.rs

```diff
@@ -0,0 +1,208 @@
+use num_bigint::BigInt;
+use num_integer::Integer;
+use num_traits::{One, Zero};
+use std::fmt;
+
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+struct Rational {
+    nominator: BigInt,
+    denominator: BigInt,
+}
+
+impl Rational {
+    fn new(n: &BigInt, d: &BigInt) -> Rational {
+        assert!(!d.is_zero(), "denominator cannot be 0");
+        // simplify if possible
+        let c = n.gcd(d);
+        Rational {
+            nominator: n / &c,
+            denominator: d / &c,
+        }
+    }
+
+    fn is_proper(&self) -> bool {
+        self.nominator < self.denominator
+    }
+    fn to_egyptian(&self) -> VecRational {
+        let mut frac: VecRational = VecRational(Vec::new());
+
+        let mut current: Rational;
+        if !self.is_proper() {
+            // input is grater than 1
+            // store the integer part
+            frac.0.push(Rational::new(
+                &self.nominator.div_floor(&self.denominator),
+                &One::one(),
+            ));
+
+            // calculate the remainder
+            current = Rational::new(
+                &self.nominator.mod_floor(&self.denominator),
+                &self.denominator,
+            );
+        } else {
+            current = self.clone();
+        }
+
+        while !current.nominator.is_one() {
+            let div = current.denominator.div_ceil(&current.nominator);
+
+            // store the term
+            frac.0.push(Rational::new(&One::one(), &div));
+
+            current = Rational::new(
+                &(-&current.denominator).mod_floor(&current.nominator),
+                match current.denominator.checked_mul(&div).as_ref() {
+                    Some(r) => r,
+                    _ => break,
+                },
+            );
+        }
+
+        frac.0.push(current);
+        frac
+    }
+}
+
+impl fmt::Display for Rational {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.denominator.is_one() {
+            // for integers only display the integer part
+            write!(f, "{}", self.nominator)
+        } else {
+            write!(f, "{}/{}", self.nominator, self.denominator)
+        }
+    }
+}
+
+#[derive(Debug)]
+struct VecRational(Vec<Rational>);
```

Comment:
I think using `Vec<Rational>` directly would be better, with a separate function for displaying it.

Reply:
Make sense. Thank you for the feedback!